packages/core/src/crawlers/context_pipeline.ts

```diff
@@ -0,0 +1,170 @@
+import type { Awaitable } from '@crawlee/types';
+
+import {
+    ContextPipelineCleanupError,
+    ContextPipelineInitializationError,
+    ContextPipelineInterruptedError,
+    RequestHandlerError,
+    SessionError,
+} from '../errors.js';
+
+/**
+ * Represents a middleware step in the context pipeline.
+ *
+ * @template TCrawlingContext - The input context type for this middleware
+ * @template TCrawlingContextExtension - The enhanced output context type
+ */
+export interface ContextMiddleware<TCrawlingContext extends {}, TCrawlingContextExtension extends {}> {
+    /** The main middleware function that enhances the context */
+    action: (context: TCrawlingContext) => Promise<TCrawlingContextExtension>;
+    /** Optional cleanup function called after the consumer finishes or fails */
+    cleanup?: (context: TCrawlingContext & TCrawlingContextExtension, error?: unknown) => Promise<void>;
```

Comment:
Let's allow sync callbacks too

```suggestion
    action: (context: TCrawlingContext) => Awaitable<TCrawlingContextExtension>;
    /** Optional cleanup function called after the consumer finishes or fails */
    cleanup?: (context: TCrawlingContext & TCrawlingContextExtension, error?: unknown) => Awaitable<void>;
```